test/unit/test.discovery.v1.js

```diff
@@ -0,0 +1,177 @@
+
+  var noop = function() {};
+
+  // Test params
+  var service = {
+    username: 'batman',
+    password: 'bruce-wayne',
+    url: 'http://ibm.com:80',
+    version: 'v1-experimental',
+    version_date: '2016-11-07'
+  };
+
+  var service1 = {
+    password: 'bruce-wayne',
+    url: 'http://ibm.com:80',
+    version: 'v1',
+    username: 'batman'
+  };
+
+  var paths = {
+    environments : '/v1/environments',
+    environmentinfo : '/v1/environments/env-guid',
+    collections: '/v1/environments/env-guid/collections',
+    collectioninfo: '/v1/environments/env-guid/collections/col-guid',
+    configurations: '/v1/environments/env-guid/configurations',
+    configurationinfo: '/v1/environments/env-guid/configurations/config-guid',
+    delete_collection: '/v1/environments/env-guid/collections/col-guid',
+    add_document: '/v1/environments/env-guid/collections/col-guid/documents',
+    delete_document: '/v1/environments/env-guid/collections/col-guid/documents/document-guid',
+    query: '/v1/environments/env-guid/collections/col-guid/query'
+  };
+
+  before(function() {
+    nock.disableNetConnect();
+    nock(service.url)
+      .persist()
+      .post(paths.environments + '?version=' + service.version_date)
+      .reply(200, {'environment_id': 'yes'})
+      .get(paths.environmentinfo + '?version=' + service.version_date)
+      .reply(200, {'environment_id': 'info'})
+      .get(paths.collections + '?version=' + service.version_date)
+      .reply(200, {'collection_id': 'yes'})
+      .get(paths.collectioninfo + '?version=' + service.version_date)
+      .reply(200, {'collection_id': 'info'})
+      .get(paths.query + '?version=' + service.version_date)
+      .reply(200, {'query': 'yes'})
+      .post(paths.collections + '?version=' + service.version_date)
+      .reply(200, {'collection_id': 'yes'})
+      .delete(paths.delete_collection + '?version=' + service.version_date)
+      .reply(200, {'config': 'yes'})
+      .post(paths.add_document + '?version=' + service.version_date)
+      .reply(200, {'add_doc': 'yes'})
+      .delete(paths.delete_document + '?version=' + service.version_date)
+      .reply(200, {'delete_doc': 'yes'})
+      .get(paths.configurations + '?version=' + service.version_date)
+      .reply(200, {'configs': 'yes'});
+  });
+
+  after(function() {
+    nock.cleanAll();
+  });
+
+  var missingParameter = function(err) {
+    assert.ok((err instanceof Error) && /required parameters/.test(err));
+  };
+
+  var discovery = new Discovery(service);
+
+  describe('discovery()', function() {
+    it('should generate a valid payload', function() {
+      var req = discovery.getEnvironments({}, noop);
+      assert.equal(req.uri.href, service.url + paths.environments + '?version=' + service.version_date);
+      assert.equal(req.method, 'GET');
+    });
+
+    it('should check no parameters provided (negative test)', function() {
+      discovery.getEnvironments({}, missingParameter);
```

Comment:
This test doesn't make any sense. It passes regardless of weather or not getEnvironments() throws a missingParameter error. (Because the test is sync, but if getEnvironments doesn't throw an error, then fires the callback async - so the test has already passed by the time `missingParameter` throws). I think we should probably make the callback always async, and make `missingParameter` accept a callback.